os/include/tinyara/sched.h

```diff
@@ -626,14 +631,18 @@ struct tcb_s {
 
 	struct xcptcontext xcp;		/* Interrupt register save area        */
 
-#ifdef CONFIG_APP_BINARY_SEPARATION
+#if defined(CONFIG_APP_BINARY_SEPARATION)
 	uint32_t ram_start;		/* Start address of RAM partition for this app */
 	uint32_t ram_size;		/* Size of RAM partition for this app */
 	uint32_t uspace;		/* User space object for app binary */
 	uint32_t uheap;			/* User heap object pointer */
+#endif
 
-#ifdef CONFIG_ARM_MPU
-	uint32_t mpu_regs[3 * MPU_NUM_REGIONS];	/* We need 3 register values to configure each MPU region */
+#if defined(CONFIG_ARM_MPU) && (defined(CONFIG_APP_BINARY_SEPARATION) || defined(CONFIG_MPU_STACKGUARD))
+#ifdef CONFIG_MPU_STACKGUARD
+	uint32_t mpu_regs[3 * (STACKGUARD_MPU_NUM + 1)]; /* Using 4 mpu regions,as stackguard region is fixed temporarily*/
```

Comment:
When we configure as below,
1. CONFIG_ARM_MPU and CONFIG_MPU_STACKGUARD are enabled
2. CONFIG_APP_BINARY_SEPARATION is disabled

Do we use 3 mpu regions regardless of mpu stackguard?
Oh, that's why you add calling `mpu_configure_app_regs` function at up_create_stack function. 
But is there any pros using 3 MPU region with protected build without APP_BINARY_SEPARATION?

Reply:
in the case mentioned above it was protected build enabled, we use total 3 MPU regions including stack guard memory protection region (1.mpu_userflash, 2.mpu_userintsram, 3.stack guard mpu region).
The region number for stack guard is fixed now in this PR. 
When there is dynamic mpu regions support is added, it will not be hard coded and this part of structure will change.

Reply:
> in the case mentioned above it was protected build enabled, we use total 3 MPU regions including stack guard memory protection region (1.mpu_userflash, 2.mpu_userintsram, 3.stack guard mpu region).

There is no setting of `1.mpu_userflash, 2.mpu_userintsram` for protected build. And I don't know that is necessary. So to use the stack guard region, it spends much more memory. This must be modified when you change it as dynamic as you said.